homeassistant/components/utility_meter/sensor.py

```diff
@@ -355,53 +377,70 @@ def start(self, unit):
         self._state = 0
         self.async_write_ha_state()
 
+    @staticmethod
+    def _validate_state(state: State | None) -> Decimal | None:
+        """Parse the state as a Decimal if available. Throws DecimalException if the state is not a number."""
+        return (
+            None
+            if state is None or state.state in [STATE_UNAVAILABLE, STATE_UNKNOWN]
+            else Decimal(state.state)
+        )
+
+    def calculate_adjustment(
+        self, old_state: State | None, new_state_val: Decimal
+    ) -> Decimal | None:
+        """Calculate the adjustment based on the old and (assumed valid) new state."""
+
+        if self._sensor_delta_values:
+            return new_state_val
+
+        if (
+            not self._sensor_periodically_resetting
+            and self._last_valid_state is not None
+        ):  # Fallback to old_state if sensor is periodically resetting but last_valid_state is None
+            return new_state_val - self._last_valid_state
+
+        try:
+            if (old_state_val := self._validate_state(old_state)) is not None:
+                return new_state_val - old_state_val
+        except DecimalException as err:
+            _LOGGER.warning(
+                "Invalid state (%s > %s): %s", old_state.state, new_state_val, err  # type: ignore[union-attr]  # because Exception only raised if old_state is not None
+            )
+        return None
+
     @callback
-    def async_reading(self, event):
+    def async_reading(self, event: Event):
         """Handle the sensor state changes."""
-        old_state = event.data.get("old_state")
-        new_state = event.data.get("new_state")
+        old_state: State | None = event.data.get("old_state")
+        new_state: State | None = event.data.get("new_state")
 
-        if self._state is None and new_state.state:
+        if new_state and self._state is None:
```

Comment:
shouldn't it be `new_state.state` ?

and why change this line ?

Reply:
The current `if ... and new_state.state` checks the state of the new_state State object whether it is None or not. However, the state attribute of a State object is always a string and therefore never None. If the actual reason for this line is to check if there is a valid new state value, we should maybe move lines 426-436 up, because that line is actually checking if the state value is valid.

I thought that the check was just there to verify if there is a new state in the Event and if so, start the meters (even when the state itself is invalid). Just let me know what is should be. I think that there is something to say for both situations, given that the result is the same.